src/core/unit.c

```diff
@@ -4280,7 +4280,18 @@ int unit_patch_contexts(Unit *u) {
                          * UID/GID around in the file system or on IPC objects. Hence enforce a strict
                          * sandbox. */
 
-                        ec->private_tmp = true;
+                        /* Enforce private tmp directories, either via TemporaryFileSystem= or PrivateTmp=,
+                         * as they have different semantics but are both enough for sanboxing purposes. */
+                        bool tmp_found = false, var_tmp_found = false;
+                        FOREACH_ARRAY(f, ec->temporary_filesystems, ec->n_temporary_filesystems)
+                                if (path_equal(f->path, "/tmp"))
+                                        tmp_found = true;
+                                else if (PATH_IN_SET(f->path, "/var", "/var/tmp"))
+                                        var_tmp_found = true;
+
+                        if (!tmp_found || !var_tmp_found)
+                                ec->private_tmp = true;
```

Comment:
this really is the wrong place.

the namespace logic deal with suppressing redundant mounts, and please keep this there. i.e. the MountMode enum dictates an ordering of mounts and their relevants, so that underlying ones are suppressed if multiple shall be applied.

right now afaics PrivateTmp= trumps TemporaryFileSystem= unconditionally. I think it would make sense to tweak this in case dynamicuser is used, and in that case put privattmp much later, after temporaryfilesystem. i.e. introduce MOUNT_PRIVATE_TMP_DYNAMIC_USER or so and put that somewhere after MOUNT_TMP

i think it makes sense to usually let PrivateTmp= win, but not if dyanmicuser=1 is set, in which case the order should be reverted.

Reply:
But this is where it's being added now? The problem is not that it's redundant, but that this flag gets enabled implicitly here...

Reply:
i think enabling the flag is fine, but i think the effect should be a slightly different one

Reply:
but see my other comment...